pragmas: rework processing and application

zerbina opened the pull request:
## Summary

Processing and application of pragmas during the first semantic
analysis pass was all done by a single routine used for all three
contexts (declarations, statements, blocks), leading to complex
code with lots of conditional logic.


The rework is meant to make future changes to pragmas easier and the
code better to understand, while also adjusting or streamlining the
semantics of some pragmas.

### General changes

- implicit pragmas (i.e. those passed to the `.push` pragma) no longer
  affect pragma statements and pragma blocks
- pragma lists in on identdefs that become empty after macro/template
  pragma application are kept, instead of being replaced with an
  `nkEmpty` node
- allow custom pragmas on iterators (both `.closure` and `.inline`)

### Pragma changes

- the `.linksys` (unused) and `.unchecked` (redundant) pragmas are
  removed
- the `.deprecated` statement's message can now be supplied by a
  constant expression
- `.cast` pragmas cannot be `push`ed anymore
- note-setting pragmas (e.g. `.hint[X]: on`) are now only allowed in
  pragma statements
- `.pure` is now disallowed in pragma statements. It could previously
  be used to mark the enclosing entity (routine or module) as `pure`
- a pushed `.dynlib` now only applies to routines (except templates)
- allow `.stackTrace` and `.lineTrace` on iterators

## Details

For pragmas on declarations and for blocks, analysis and application
are separated into two sequential steps, where first all pragmas are
analysed (`.pragma` pragmas expanded, custom pragmas resolved, early
rejection of unapplicable pragmas) and then all non-erroneous ones
applied.

Since a single pragma in statement can alter note state (which could
affect analysis of the pragmas following it), analysis and application
for statement pragmas happens in an intertwined manner.

Handling of the three special pragmas `.pragma`, `.push`, and `.pop`
no longer relies on a node flag (`nfImplicitPragma`, which is now
removed), but instead happens separately from the other pragmas.

In order to remove the extra arguments previously passed to the
`pragma` procedure, each context where pragmas are applied now uses its
own dedicated procedure.

In general, `nkError` node propagation is improved and the amount of
input AST mutations reduced.

---

## To-do
- [x] apply implicit pragmas to all declarations again
- [x] wrap errors occurring during application of pushed pragmas (i.e. implicit pragmas) in `adSemImplicitPragmaError`s
- [x] use more consistent naming
- [x] write a commit message

## Notes for reviewers
- the goal is to make further changes/improvements to pragma easier and the code easier to navigate

saem commented on compiler/sem/pragmas.nim:
```diff
@@ -129,6 +129,10 @@ const
   allRoutinePragmas* = methodPragmas + iteratorPragmas + lambdaPragmas
   enumFieldPragmas* = {wDeprecated}
 
+  allowsCustomPragma = {skVar, skLet, skParam, skField, skProc, skFunc,
```
shouldn't this include iterator?

zerbina replied:
I don't see a reason why not. The set was just copied from the previous the code, without me adjusting it.

zerbina replied:
Custom pragmas are now also allowed on iterators.

saem commented on compiler/sem/pragmas.nim:
```diff
@@ -1563,90 +1377,56 @@ proc prepareSinglePragma(
-        else:
+          # XXX: modifying the module graph during application of a symbol
+          #      operator doesn't seem like a good idea...
           result = magicsys.registerNimScriptSymbol2(c.graph, sym)
```
This is clean-up I forgot, we can drop the old `registerNimScriptSymbol` and take over the name. Doesn't have to be tackled in this PR, I can do it after.

saem commented on compiler/sem/pragmas.nim:
```diff
@@ -1563,90 +1377,56 @@ proc prepareSinglePragma(
-        if sym == nil:
-          result = invalidPragma(c, it)
-        else:
+          # XXX: modifying the module graph during application of a symbol
```
At first I thought it was in effect export it, but nope, it's adding it to `exposed`, which I believe amounts to export it at the codegen level, but not the semantic level. I can't seem to find any usage of `exportnims` either -- odd.

zerbina replied:
I believe it's meant for registering a procedure to later be able to look it up by name, when embedding NimScript.

saem replied:
Yup, at the time I made the comment I did some searches trying to find the mechanism where things were actually stitched together. It was really difficult to find and couldn't get a complete picture. With that said, if that specific bit is revisited, that whole mechanism likely needs to be revised and perhaps this pragma will be shown to be unnecessary.

Not expecting any additional action.

saem commented on compiler/sem/semtypes.nim:
```diff
@@ -56,7 +56,7 @@ proc semEnum(c: PContext, n: PNode, prev: PType): PType =
-        n[i][0][1] = pragma(c, e, n[i][0][1], enumFieldPragmas)
+        n[i][0][1] = pragmaDecl(c, e, n[i][0][1], enumFieldPragmas)
         # check if we got any errors and if so report them
         for e in ifErrorWalkErrors(c.config, n[i][0][1]):
```
I reallly like the direction of the PR. I think once these changes are in, I can swap out these walk errors with propagating the errors instead. That should clean-up a bunch of downstream code. 🎉

saem commented on compiler/ast/ast_types.nim:
```diff
@@ -591,13 +591,6 @@ type
     nfDefaultRefsParam ## a default param value references another parameter
                        ## the flag is applied to proc default values and to calls
     nfHasComment ## node has a comment
-    nfImplicitPragma ## node is a "singlePragma" this is a transition flag
```
YES!

saem commented on compiler/front/cli_reporter.nim:
```diff
@@ -1640,7 +1637,7 @@ proc reportBody*(conf: ConfigRef, r: SemReport): string =
 
     of rsemImplicitPragmaError:
-      result = "???"
+      result = "application of implicit pragma failed"
```
nice catch!

saem commented on compiler/sem/pragmas.nim:
```diff
@@ -572,37 +574,42 @@ proc processOption(c: PContext, n: PNode, resOptions: var TOptions): PNode =
-  if n[start-1].kind in nkPragmaCallKinds:
-    result = c.config.newError(n, PAstDiag(kind: adSemUnexpectedPushArgument))
-    return
+proc processPush(c: PContext, info: TLineInfo, pragmas: openArray[PNode], r: PNode): bool =
```
Might be too big a change, but return `true` on success "feels" better.

saem commented on compiler/sem/pragmas.nim:
```diff
@@ -572,37 +574,42 @@ proc processOption(c: PContext, n: PNode, resOptions: var TOptions): PNode =
-      assert not cyclicTree(result)
+      r.add err
+      result = true
       return
```
This is likely a bug on my part, should probably continue even if there are errors (for check, suggest, etc). Don't have to fix now.

saem commented on compiler/sem/pragmas.nim:
```diff
@@ -835,18 +842,20 @@ proc pragmaLine(c: PContext, n: PNode): PNode =
-    n[i] = invalidPragma(c, it)
-    result = n[i]
-    return
+    return invalidPragma(c, it)
```
This was early `nkError` conversion, I could have done this better; this doesn't need to be changed now.

What should happen is the error should encompass `it` and the rest of the nodes -- not just `it`. Doing it that way means we can recover/restore from `nkError` by grabbing `wrongNode` from the diag.

saem commented on compiler/sem/pragmas.nim:
```diff
@@ -1044,179 +1026,94 @@ proc processEffectsOf(c: PContext, n: PNode; owner: PSym): PNode =
+        assert sym.kind == skTemplate
+        incl(sym.flags, sfDirty)
       of wImportJs:
+        # note: don't use ``semExternName`` here. The operand is *not* an
```
Hmm, at a later point we likely need separate `importjs` pragmas, because we have the whole calling/emit pattern language hidding in here.

saem commented on compiler/sem/pragmas.nim:
```diff
@@ -1269,10 +1162,11 @@ proc prepareSinglePragma(
         incl(sym.flags, sfNoalias)
       of wEffectsOf:
         result = processEffectsOf(c, it, sym)
+        if result.isNil: # was there no error?
```
`processEffectsOf` and `processParam` therein are messing this up by silently dropping `semExpr` errors, and they really should create a full production. That doesn't have to be done now; pre-existing issues.

saem commented on compiler/sem/pragmas.nim:
```diff
@@ -1361,29 +1249,27 @@ proc prepareSinglePragma(
-        if sym.typ == nil:
-          result = invalidPragma(c, it)
-        else: incl(sym.typ.flags, tfFinal)
+        assert sym.typ != nil
```
(pre-existing issue) This should do a reciprocal check of `wInheritable` in case one pragma is applied before the other.

zerbina replied:
Applying `wFinal` when the type is already marked as `.inheritable` is now an error.

saem commented on compiler/sem/pragmas.nim:
```diff
@@ -1438,119 +1299,53 @@ proc prepareSinglePragma(
-      of wPop:
-        result = processPop(c, it)
-        result.flags.incl nfImplicitPragma # xxx: legacy singlepragma=true
+          # XXX: disallow for the other declarations (e.g. ``var``)?
```
yeah, that's weird, because `wError` is in `declPragmas` whcih is under `varPragmas`. 🤔 I bet the rest of the compiler can't handle it.

zerbina replied:
I restored the previous behaviour (`.error`s on, for example, a `var` being treated like they are in statements) for now.

saem commented on compiler/sem/pragmas.nim:
```diff
@@ -1438,119 +1299,53 @@ proc prepareSinglePragma(
-          var tmp = c.config.options
-          result = processOption(c, it, tmp)
-          c.config.options = tmp
+        # TODO: also support for funcs and (closure) iterators -- there's no
+        #       reason not to
```
Strongly agree.

saem commented on compiler/sem/pragmas.nim:
```diff
@@ -1438,119 +1299,53 @@ proc prepareSinglePragma(
+        assert sym.typ != nil
+        incl(sym.typ.flags, tfCompleteStruct)
       of wUnchecked:
+        # XXX: seems redundant with ``tyUncheckedArray``. Remove it?
```
Yup, it was supposedly removed in Nim 0.20.0 (per the change log).

zerbina replied:
Removed the pragma and special word.

saem reviewed (commented):
I'm part way done, going to `processBlockPragma` next.

saem commented on compiler/sem/pragmas.nim:
```diff
@@ -1657,65 +1418,324 @@ proc prepareSinglePragma(
+  case k
+  of wUsed:
+    result = noVal(c, it)
+    # XXX: an escaping mutation if the the pragma is used inside a ``compiles``
+    #      block
```
```suggestion
    # XXX: an escaping mutation if the pragma is used inside a ``compiles``
    #      block
```

saem replied:
🤔 Future thought, if we had a more general `recordPragma` facility we could record all pragma application events, then unwind them when leaving compile or other such speculative/sandboxed execution contexts.

zerbina replied:
Yep, I agree. Specifically for the `wUsed` pragma in *statement* pragmas, disallowing it might be better, I think. The following is currently allowed and works:
```nim
proc p() =
  var x = 0
  {.used.} # `p` is marked as used
```
which I'm not sure is a good idea.

saem replied:
Yup, that's not a good idea, it'll be easier for something to remove after this refactor! 🚀

saem commented on compiler/sem/pragmas.nim:
```diff
@@ -1657,65 +1418,324 @@ proc prepareSinglePragma(
+    owner.flags.incl sfUsed
+  of wDeprecated:
+    result = it
+    # TODO: disallow when the owner is not a module
```
Wouldn't this preclude using deprecated for example on inner routines? It's a handy trick when working with really large proc refactoring.

zerbina replied:
It wouldn't, no. The `wDeprecated` handling here is only for pragma statements, which I believe were only intended to be used with modules, but due to the previous application architecture were (and currently still are) also usable inside procedures, like so:
```nim
proc p() =
  var x = 0
  {.deprecated.} # previously applied to `p`, but now to the module `p` is part of as deprecated
```

I think this usage should be disallowed, hence the todo.

**Edit**: corrected the description of how `.deprecated` works when in a statement.

saem replied:
Got it, thanks for the explanation.

saem commented on compiler/sem/pragmas.nim:
```diff
@@ -1657,65 +1418,324 @@ proc prepareSinglePragma(
+    if not isTopLevel(c):
+      result = c.config.newError(it,
+                  PAstDiag(kind: adSemExperimentalRequiresToplevel))
+    result = processExperimental(c, it)
```
Not sure what was going on with the old code, but this should be if/else or maybe:

```suggestion
    result = processExperimental(c, it)
    if not isTopLevel(c):
      result = c.config.newError(result,
                  PAstDiag(kind: adSemExperimentalRequiresToplevel))

```

zerbina replied:
An `if`/`else` is indeed better, thanks.

saem commented on compiler/sem/pragmas.nim:
```diff
@@ -1657,65 +1418,324 @@ proc prepareSinglePragma(
+      result = c.config.newError(it,
+                  PAstDiag(kind: adSemExperimentalRequiresToplevel))
+    result = processExperimental(c, it)
+  else:
```
I was about to say that this switch should be exhaustive, but then I did an inventory of pragmas and here is what I got:

The constant `stmtPragmas` is missing the following pragmas from this case statement:
- `wCallConv` (same as pushing a calling convention, cooooool)
- `wDebugger` (dead with the embedded nim debugger, noop explicitly in the case?)
- `wProfiler`
- `wByRef`

The constant `stmtPragmas` has the following that are missing from this case statement:
- `wLinksys`
- `wPure` 
- `wDeadCodeElimUnused` (this one is deprecated per the comment)
- `wEffects` (pragma remains and `sempass2` sees it and emits effects at that point)
- `wAssert` (does nothing, might have been an old way to force an assertion)

The first set can be safely added to `stmtPragmas`. For the second set I would add `wEffects` to the case, I don't think it belongs in the symbol pragmas, the rest should be safe to drop from the constant. At which point this case can be exhaustive.

zerbina replied:
I made the case statement exhaustive, although with two differences:
- `wDeadCodeElimUnused` is kept as a no-op, with a comment suggesting its removal
- `wAssert` is kept -- it can be used to enable/disable assertions for a section of code

zerbina replied:
Oh, I misread the code, `wAssert` is indeed unused.

saem commented on compiler/sem/pragmas.nim:
```diff
@@ -1438,119 +1299,53 @@ proc prepareSinglePragma(
-        result = noVal(c, it)
+        sym.typ.callConv = wordToCallConv(k)
+        sym.typ.flags.incl tfExplicitCallConv
       of wEffects:
```
I think this needs to move to the `applyStmtPragma`, I don't believe it's a symbol pragma.

zerbina replied:
You're very much correct. I'm also going to add a test for the pragma, as the testsuite doesn't seem to have caught this mistake.

saem commented on compiler/sem/pragmas.nim:
```diff
@@ -1657,65 +1418,324 @@ proc prepareSinglePragma(
+
+proc semSinglePragmaInStmt(
+    c: PContext; owner: PSym, it: PNode, validPragmas: TSpecialWords, r: var seq[PNode]): TSpecialWord =
+  ## Processes a the single pragma expression `it`. If the pragma is a user-
```
```suggestion
  ## Processes the single pragma expression `it`. If the pragma is a user-
```

saem commented on compiler/sem/pragmas.nim:
```diff
@@ -1728,60 +1748,171 @@ proc hasPragma*(n: PNode, pragma: TSpecialWord): bool =
+  ## by sem in the context of applying pragmas to an expression.
+  assert n.kind == nkPragma
+  result = copyNode(n)
+  for i in 0..<n.len:
```
Just curious why favour indexing over item iteration?

zerbina replied:
I don't know :P. The body likely required access to the index at one point, but there's no reason to not use item iteration now.

saem commented on compiler/sem/pragmas.nim:
```diff
@@ -1728,60 +1748,171 @@ proc hasPragma*(n: PNode, pragma: TSpecialWord): bool =
+      break
+    of wPop:
+      result.add processPop(c, n[i])
+      # XXX: maybe it'd be better to either report an error if there are items
```
Hmm, on module close if there are unpopped pragmas that's an error? I think that's a good idea. Basically someone forgot to "clean-up" and this would be especially helpful if someone is stuck with push/pop across conditional compilation.

zerbina replied:
The comment refers to all pragmas following a `pop` being silently dropped (e.g. `{.pop, error.}` doesn't abort compilation with a user error). This mirrors the previous behaviour, but I think in the future, the compiler should produce an error (or at least a warning) instead.

Thinking about it more, the three special pragmas should probably also be enforced to also come first (but not as part of this PR).

saem replied:
I think the idea at some point was to allow "filtered" pop, so a pop would only pop the items that followed it. This unstructured popping, seems... dangerous.

I agree we should produce an error in the future.

saem reviewed (commented):
All done; I think `wEffects` handling is the only error I noticed.

Rest are not blockers, but the exhaustiveness on case statements would be nice if feasible. If not, it's already such a big improvement. 🤩

zerbina committed:
pragmas: rework processing and application

Summary
=======

Processing and application of pragmas during the first semantic
analysis pass was all done by a single routine used for all three
contexts (declarations, statements, blocks), leading to complex
code with lots of conditional logic.

The rework is meant to make future changes to pragmas easier and the
code better to understand, while also adjusting or streamlining the
semantics of some pragmas.

General changes
---------------

- implicit pragmas (i.e. those passed to the `.push` pragma) no longer
  affect pragma statements and pragma blocks
- pragma lists in on identdefs that become empty after macro/template
  pragma application are kept, instead of being replaced with an
  `nkEmpty` node
- allow custom pragmas on iterators (both `.closure` and `.inline`)

Pragma changes
--------------

- the `.linksys` (unused), `.assert` (unused), and `.unchecked`
  (redundant) pragmas are removed
- the `.deprecated` statement's message can now be supplied by a
  constant expression
- `.cast` pragmas cannot be `push`ed anymore
- note-setting pragmas (e.g. `.hint[X]: on`) are now only allowed in
  pragma statements
- `.pure` is now disallowed in pragma statements. It could previously
  be used to mark the enclosing entity (routine or module) as `pure`
- a pushed `.dynlib` now only applies to routines (except templates)
- allow `.stackTrace` and `.lineTrace` on iterators

Details
=======

For pragmas on declarations and for blocks, analysis and application
are separated into two sequential steps, where first all pragmas are
analysed (`.pragma` pragmas expanded, custom pragmas resolved, early
rejection of unapplicable pragmas) and then all non-erroneous ones
applied.

Since a single pragma in statement can alter note state (which could
affect analysis of the pragmas following it), analysis and application
for statement pragmas happens in an intertwined manner.

Handling of the three special pragmas `.pragma`, `.push`, and `.pop`
no longer relies on a node flag (`nfImplicitPragma`, which is now
removed), but instead happens separately from the other pragmas.

In order to remove the extra arguments previously passed to the
`pragma` procedure, each context where pragmas are applied now uses its
own dedicated procedure.

In general, `nkError` node propagation is improved and the amount of
input AST mutations reduced.

saem commented:
Did a scan of the comments and recent changes, seems like anything that would remotely block this PR are done -- cancel if not.

Take it away,
bors r+
